staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator/aggregator.go

```diff
@@ -17,331 +17,240 @@ limitations under the License.
-		}
-		if err := aggregator.MergeSpecsIgnorePathConflictRenamingDefinitionsAndParameters(specToReturn, specInfo.spec); err != nil {
-			return nil, err
-		}
-	}
-
-	return specToReturn, nil
+	info.spec.Replace(spec)
+	s.specsByAPIServiceName[name] = info
 }
 
-// updateOpenAPISpec aggregates all OpenAPI specs.  It is not thread-safe. The caller is responsible to hold proper locks.
-func (s *specAggregator) updateOpenAPISpec() error {
-	if s.openAPIVersionedService == nil {
-		return nil
+// buildMergeSpecLocked creates a new cached mergeSpec from the list of cached specs.
+func (s *specAggregator) buildMergeSpecLocked() cached.Data[*spec.Swagger] {
+	apiServices := make([]*v1.APIService, 0, len(s.specsByAPIServiceName))
+	for k := range s.specsByAPIServiceName {
+		apiServices = append(apiServices, &s.specsByAPIServiceName[k].apiService)
 	}
-	specToServe, err := s.buildOpenAPISpec()
-	if err != nil {
-		return err
-	}
-	return s.openAPIVersionedService.UpdateSpec(specToServe)
-}
-
-// tryUpdatingServiceSpecs tries updating openAPISpecs map with specified specInfo, and keeps the map intact
-// if the update fails.
-func (s *specAggregator) tryUpdatingServiceSpecs(specInfo *openAPISpecInfo) error {
-	if specInfo == nil {
-		return fmt.Errorf("invalid input: specInfo must be non-nil")
+	sortByPriority(apiServices)
+	caches := make([]cached.Data[*spec.Swagger], len(apiServices))
+	for i, apiService := range apiServices {
+		caches[i] = &(s.specsByAPIServiceName[apiService.Name].spec)
 	}
-	_, updated := s.openAPISpecs[specInfo.apiService.Name]
-	origSpecInfo, existedBefore := s.openAPISpecs[specInfo.apiService.Name]
-	s.openAPISpecs[specInfo.apiService.Name] = specInfo
 
-	// Skip aggregation if OpenAPI spec didn't change
-	if existedBefore && origSpecInfo != nil && origSpecInfo.etag == specInfo.etag {
-		return nil
-	}
-	klog.V(2).Infof("Updating OpenAPI spec because %s is updated", specInfo.apiService.Name)
-	defer func(start time.Time) {
-		duration := time.Since(start)
-		klog.V(2).Infof("Finished OpenAPI spec generation after %v", duration)
-
-		reason := "add"
-		if updated {
-			reason = "update"
-		}
-
-		regenerationCounter.With(map[string]string{"apiservice": specInfo.apiService.Name, "reason": reason})
-		regenerationDurationGauge.With(map[string]string{"reason": reason}).Set(duration.Seconds())
-	}(time.Now())
-	if err := s.updateOpenAPISpec(); err != nil {
-		if existedBefore {
-			s.openAPISpecs[specInfo.apiService.Name] = origSpecInfo
-		} else {
-			delete(s.openAPISpecs, specInfo.apiService.Name)
+	return cached.NewListMerger(func(results []cached.Result[*spec.Swagger]) cached.Result[*spec.Swagger] {
+		var merged *spec.Swagger
+		etags := make([]string, 0, len(results))
+		for _, specInfo := range results {
+			result := specInfo.Get()
+			if result.Err != nil {
+				// APIService name and err message will be included in
+				// the error message as part of decorateError
+				klog.Warning(result.Err)
```

Comment:
for a follow-up: does this error have context, i.e. the apiservice name?

Reply:
Yes (ref: https://github.com/kubernetes/kubernetes/blob/ea23e13463f48b788b9c9d0ef0d88a4ee77deda7/staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator/aggregator.go#L252), also in the comments above:

```
// APIService name and err message will be included in
// the error message as part of decorateError
```

Reply:
ha, missed that comment :D